dmoj/cptbox/isolate.py

```diff
@@ -201,176 +202,290 @@ def __init__(self, read_fs, write_fs=None, writable=(1, 2)):
                 return True
 
-            file, error = self._file_access_check(file, debugger, fs_jail(debugger), dirfd=debugger.arg0)
-            if not error:
-                return True
+            dirfd = self.get_reg_value(debugger, dir_reg)
+            full_path = self.get_full_path(debugger, rel_file, dirfd)
+            return self._raw_access_check(syscall, debugger, full_path, self.read_fs_jail)
+
+        return check
+
+    @staticmethod
+    def wrap_access_check_dec(syscall: str) -> Callable[[AccessChecker], HandlerCallback]:
+        def inner(func: AccessChecker) -> HandlerCallback:
+            return IsolateTracer.wrap_access_check(syscall, func)
+
+        return inner
+
+    @staticmethod
+    def wrap_access_check(syscall: str, access_check: AccessChecker) -> HandlerCallback:
+        def inner(debugger) -> bool:
+            try:
+                allowed = access_check(debugger)
+            except AccessFailure as failure:
+                failure.log_failure(syscall)
+                return failure.handle(debugger)
+            else:
+                if not allowed:
+                    return ACCESS_EACCES(debugger)
+                else:
+                    return True
+
+        return inner
+
+    DirFDGetter = Callable[[Debugger], int]
+
+    def _dirfd_getter_from_reg(self, reg: int) -> DirFDGetter:
+        def getter(debugger: Debugger) -> int:
+            return self.get_reg_value(debugger, reg)
+
+        return getter
 
-            log.debug('Denied access via syscall %s (error: %s): %s', syscall, error.error_name, file)
-            return error(debugger)
+    def _dirfd_getter_cwd(self, debugger: Debugger) -> int:
+        return AT_FDCWD
+
+    FSJailGetter = Callable[[Debugger], FilesystemPolicy]
+
+    def _fs_jail_getter_from_open_flags_reg(self, reg: int) -> FSJailGetter:
+        def getter(debugger: Debugger) -> FilesystemPolicy:
+            open_flags = self.get_reg_value(debugger, reg)
+            return self._fs_jail_from_open_flags(open_flags)
+
+        return getter
+
+    def _fs_jail_getter_from_kind(self, kind) -> FSJailGetter:
+        fs_jail = self._fs_jail_from_kind(kind)
+        return lambda _: fs_jail
+
+    def _fs_jail_from_open_flags(self, open_flags: int) -> FilesystemPolicy:
+        for flag in open_write_flags:
+            # Strict equality is necessary here, since e.g. O_TMPFILE has multiple bits set,
+            # and O_DIRECTORY & O_TMPFILE > 0.
+            if open_flags & flag == flag:
+                return self.write_fs_jail
+
+        return self.read_fs_jail
+
+    def _fs_jail_from_kind(self, kind: FilesystemSyscallKind) -> FilesystemPolicy:
+        if kind == FilesystemSyscallKind.READ:
+            return self.read_fs_jail
+        else:
+            assert kind == FilesystemSyscallKind.WRITE, 'Bad FilesystemSyscallKind'
+            return self.write_fs_jail
+
+    def main_access_check(
```

Comment:
"main", as opposed to...? This is the only "public" access check function, feel free to call it `access_check` and the internal one just `_access_check`.